File: src/jvm/main/org/jetbrains/kotlinx/lincheck/verifier/CachedVerifier.java
/*
 * Lincheck
 *
 * Copyright (C) 2019 - 2023 JetBrains s.r.o.
 *
 * This Source Code Form is subject to the terms of the
 * Mozilla Public License, v. 2.0. If a copy of the MPL was not distributed
 * with this file, You can obtain one at http://mozilla.org/MPL/2.0/.
 */

package org.jetbrains.kotlinx.lincheck.verifier;

import org.jetbrains.kotlinx.lincheck.execution.*;

import java.util.*;

/**
 * This verifier cached the already verified results in a hash table,
 * and look into this hash table at first. In case of many invocations
 * with the same scenario, this optimization improves the verification
 * phase significantly.
 */
public abstract class CachedVerifier implements Verifier {
    private final Map<ExecutionScenario, Set<ExecutionResult>> previousResults = new WeakHashMap<>();

    @Override
    public boolean verifyResults(ExecutionScenario scenario, ExecutionResult results) {
        Set<ExecutionResult> executionResults = previousResults.computeIfAbsent(scenario, s -> new HashSet<>());
        if (executionResults.contains(results)) return true;
        boolean isValid = verifyResultsImpl(scenario, results);
        if (isValid) {
            executionResults.add(results);
        }
        return isValid;
    }

    public abstract boolean verifyResultsImpl(ExecutionScenario scenario, ExecutionResult results);
}

File: src/jvm/main/org/jetbrains/kotlinx/lincheck/verifier/CachedVerifier.kt
/*
* #%L
* Lincheck
* %%
* Copyright (C) 2015 - 2018 Devexperts, LLC
* %%
* This program is free software: you can redistribute it and/or modify
* it under the terms of the GNU Lesser General Public License as
* published by the Free Software Foundation, either version 3 of the
* License, or (at your option) any later version.
*
* This program is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
* GNU General Lesser Public License for more details.
*
* You should have received a copy of the GNU General Lesser Public
* License along with this program.  If not, see
* <http://www.gnu.org/licenses/lgpl-3.0.html>.
* #L%
*/
package org.jetbrains.kotlinx.lincheck.verifier

import org.jetbrains.kotlinx.lincheck.execution.ExecutionResult
import org.jetbrains.kotlinx.lincheck.execution.ExecutionScenario
import java.util.*
import kotlin.collections.HashSet

/**
 * This verifier cached the already verified results in a hash table,
 * and look into this hash table at first. In case of many invocations
 * with the same scenario, this optimization improves the verification
 * phase significantly.
 */
abstract class CachedVerifier : Verifier {
    private val previousResults: MutableMap<ExecutionScenario, MutableSet<ExecutionResult>> = WeakHashMap()
    override fun verifyResults(scenario: ExecutionScenario, results: ExecutionResult): Boolean {
        val newResult = previousResults.computeIfAbsent(scenario) { s: ExecutionScenario? -> HashSet() }.add(results)
        return if (!newResult) true else verifyResultsImpl(scenario, results)
    }

    abstract fun verifyResultsImpl(scenario: ExecutionScenario, results: ExecutionResult): Boolean
}